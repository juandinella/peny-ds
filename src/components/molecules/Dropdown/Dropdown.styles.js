import { css } from 'styled-components'

export default css`
  position: relative;
  height: 100%;

  .dropdown-menu {
    background: var(--color-neutral-00);
    padding: var(--spacing-xs) var(--spacing-sm);
    border-radius: var(--radius-sm);
    box-shadow: var(--shadow-md);
    padding: var(--spacing-xxs) 0;
    opacity: 0;
    visibility: hidden;
    transition: transform 0.2s, opacity 0.2s, visibility 0s linear 0.2s;
    will-change: transform;
    position: absolute;
    width: max-content;
    margin: var(--spacing-xxs) auto auto;
    z-index: 2;
    list-style: none;

    li {
      padding: var(--spacing-xs) var(--spacing-md);
      margin: 0;
      transition: all var(--transition-timing-easing);
      cursor: pointer;
      color: var(--color-neutral-120);

      &:hover {
        background-color: var(--color-neutral-20);
        color: var(--color-primary-main);

        a {
          color: var(--color-primary-main);
        }
      }

      &:active {
        color: var(--color-primary-dark);

        a {
          color: var(--color-primary-dark);
        }
      }

      a {
        text-decoration: none;
        color: var(--color-neutral-120);
        transition: all var(--transition-timing-easing);
      }
    }

    &.is-open {
      transform: translateY(-0.3rem);
      transition: transform 0.2s, opacity 0.2s, visibility linear;
      visibility: visible;
      opacity: 1;
    }
  }

  ${(props) =>
    props.maxHeight &&
    css`
      .dropdown-menu {
        height: ${(props) => `${props.maxHeight}px`};
        overflow: scroll;
      }
    `}

  .dropdown-toggle {
    min-width: max-content;

    svg {
      display: block;
      margin: 0 auto;
    }
  }

  [aria-expanded='true'] [aria-labelledby='IconKebab-label'] path {
    fill: var(--color-primary-dark);
  }

  [aria-expanded='true'] [aria-labelledby='IconChevronDown-label'] {
    transform: rotate(180deg);
  }

  [aria-labelledby='IconChevronDown-label'] {
    transition: transform 0.225s ease;
  }

  ${(props) =>
    props.alignment === 'left' &&
    css`
      .dropdown-menu {
        left: 0;
      }
    `}

  ${(props) =>
    props.alignment === 'center' &&
    css`
      .dropdown-menu {
        left: 0;
        right: 0;
      }
    `}

  ${(props) =>
    props.alignment === 'right' &&
    css`
      .dropdown-menu {
        right: 0;
      }
    `}

  ${(props) =>
    props.disabled &&
    css`
      button {
        cursor: not-allowed;
      }
    `}
`
