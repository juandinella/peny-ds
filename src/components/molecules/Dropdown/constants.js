export const options = {
  /**
   * Position of the Dropdown Menu
   * - **left**
   * - **center**
   * - **right**
   */
  alignments: ['left', 'center', 'right'],
}
