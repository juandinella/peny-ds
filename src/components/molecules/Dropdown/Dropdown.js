import React, { useState, useEffect, useRef } from 'react'
import PropTypes from 'prop-types'
import { options } from './constants'
import styles from './Dropdown.styles'
import TextList from '../../atoms/TextList'
import Skeleton from '../../atoms/Skeleton'
import Utils from '../../../utils'

/**
 * Display a dropdown list of items.
 */

const Dropdown = ({
  className,
  id,
  disabled,
  buttonContent,
  textList,
  alignment,
  shouldRender,
  maxHeight,
}) => {
  const wrapperRef = useRef()
  const [active, setActive] = useState('')

  const toggleDropdown = () => {
    setActive(active === '' ? 'is-open' : '')
  }

  const handleClickOutside = (e) => {
    if (wrapperRef.current.contains(e.target)) {
      return
    }
    setActive('')
  }

  useEffect(() => {
    document.addEventListener('mousedown', handleClickOutside)
    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [])

  return (
    <Skeleton
      id={`${id}_skeleton`}
      show={!Utils.emptyValue(shouldRender) && !shouldRender()}
    >
      <nav
        css={styles}
        className={className}
        id={id}
        data-testid={id}
        ref={wrapperRef}
        alignment={alignment}
        maxHeight={maxHeight}
        disabled={disabled}
      >
        <button
          className="dropdown-toggle"
          id={id ? `${id}_toggle` : ''}
          data-testid={id ? `${id}_toggle` : ''}
          onClick={toggleDropdown}
          aria-haspopup="true"
          aria-expanded={active ? 'true' : 'false'}
          aria-label="Dropdown Menu"
        >
          {buttonContent}
        </button>
        {textList && !disabled ? (
          <TextList
            className={`dropdown-menu ${active}`}
            id={id ? `${id}_list` : ''}
            data-testid={id ? `${id}_list` : ''}
            list={textList.map((item, index) => {
              return {
                ...item,
                id: `${id}_list_${index}`,
                onClick: () => {
                  item.onClick()
                  setActive(active === '' ? 'is-open' : '')
                },
              }
            })}
          />
        ) : null}
      </nav>
    </Skeleton>
  )
}

Dropdown.displayName = 'Dropdown'

Dropdown.propTypes = {
  /** Dropdown's id */
  id: PropTypes.string,
  /** To add style */
  className: PropTypes.string,
  /** Alignmnet dropdown menu */
  alignment: PropTypes.oneOf(options.alignments),
  /** Disabled */
  disabled: PropTypes.bool,
  /** Content */
  buttonContent: PropTypes.oneOfType([PropTypes.string, PropTypes.node]),
  /** Set maxHeight of the dropdown */
  maxHeight: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /** Array of objects for text list */
  textList: PropTypes.arrayOf(
    PropTypes.shape({
      /** Content for the list item. */
      content: PropTypes.oneOfType([PropTypes.string, PropTypes.node])
        .isRequired,
      /** Extends the styles applied to the list item.*/
      className: PropTypes.string,
      /** Callback fired when item list is clicked. */
      onClick: PropTypes.func,
      /** Nodes after list item. */
      children: PropTypes.node,
    })
  ),
  /**
   * Function that decides if the component should be rendered.
   *
   * @return {boolean} false if component should not be rendered and show skeleton instead. Component renders otherwise.
   */
  shouldRender: PropTypes.func,
}

Dropdown.defaultProps = {
  alignment: 'center',
}

export default Dropdown
