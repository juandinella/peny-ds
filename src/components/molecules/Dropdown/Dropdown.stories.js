import React from 'react'
import Dropdown from './Dropdown'
import { options } from './constants'

import {
  getTemplate,
  getListTemplate,
  getOptionsArgTypes,
} from '../../../helpers/storybook'

const Template = getTemplate(Dropdown)
const ListTemplate = getListTemplate(Dropdown)

export default {
  title: 'Molecules/Dropdown',
  component: Dropdown,
  args: {
    buttonContent: 'Toggle Dropdown',
    id: 'dropdown_id',
    textList: [
      {
        content: <a href="#">Im a Link</a>,
      },
      { content: 'Item Two' },
      { content: 'Item Three' },
    ],
  },
  argTypes: {
    alignment: getOptionsArgTypes(options.alignments),
  },
}

export const Default = Template.bind({})

export const Alignments = ListTemplate.bind({})
Alignments.args = {
  items: options.alignments.map((alignment) => ({ alignment })),
}

export const Skeleton = Template.bind({})
Skeleton.args = { shouldRender: () => false }
