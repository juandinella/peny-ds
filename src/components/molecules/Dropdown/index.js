import Dropdown from './Dropdown'

export default Dropdown
export { options } from './constants'
