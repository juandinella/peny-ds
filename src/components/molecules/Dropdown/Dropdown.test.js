import React from 'react'

import { render, cleanup, fireEvent } from '@testing-library/react'

import Dropdown from './Dropdown'

import initStoryshots from '@storybook/addon-storyshots'

initStoryshots({
  storyKindRegex: /^Molecules\/Dropdown$/,
})

const componentProps = {
  id: 'dropdown_id',
  buttonContent: 'Toggle Dropdown',
  textList: [
    {
      content: <a href="#">Im a Link</a>,
    },
    { content: 'Item Two' },
  ],
}

afterEach(cleanup)

describe('<Dropdown />', () => {
  test('should render component correctly', () => {
    const { queryByTestId } = render(<Dropdown {...componentProps} />)

    expect(queryByTestId(componentProps.id)).toBeTruthy()
  })

  test('should render skeleton', () => {
    const { queryByTestId } = render(
      <Dropdown {...componentProps} shouldRender={() => false} />
    )

    expect(queryByTestId(`${componentProps.id}_skeleton`)).toBeTruthy()
  })

  test('should show button content correctly', () => {
    const { queryByTestId } = render(<Dropdown {...componentProps} />)

    expect(queryByTestId(`${componentProps.id}_toggle`).textContent).toEqual(
      componentProps.buttonContent
    )
  })

  test('should open dropdown', () => {
    const { queryByTestId } = render(<Dropdown {...componentProps} />)

    const button = queryByTestId(`${componentProps.id}_toggle`)

    expect(button.getAttribute('aria-expanded')).toEqual('false')
    expect(queryByTestId(`${componentProps.id}_list`).className).not.toContain(
      'is-open'
    )

    fireEvent.click(button)

    expect(button.getAttribute('aria-expanded')).toEqual('true')
    expect(queryByTestId(`${componentProps.id}_list`).className).toContain(
      'is-open'
    )
  })

  test('should show list correctly', () => {
    const { queryByTestId } = render(<Dropdown {...componentProps} />)

    expect(queryByTestId(`${componentProps.id}_list_0`)).toBeTruthy()
    expect(queryByTestId(`${componentProps.id}_list_0`).textContent).toEqual(
      'Im a Link'
    )

    expect(queryByTestId(`${componentProps.id}_list_1`)).toBeTruthy()
    expect(queryByTestId(`${componentProps.id}_list_1`).textContent).toEqual(
      componentProps.textList[1].content
    )
  })

  test('should not show list if disabled', () => {
    const { queryByTestId, rerender } = render(<Dropdown {...componentProps} />)

    expect(queryByTestId(`${componentProps.id}_list`)).toBeTruthy()

    rerender(<Dropdown {...componentProps} disabled={true} />)

    expect(queryByTestId(`${componentProps.id}_list`)).toBeFalsy()
  })
})
