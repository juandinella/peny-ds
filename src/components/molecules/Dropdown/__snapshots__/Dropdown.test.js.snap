// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Molecules/Dropdown Alignments 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <nav
    className="sc-sg24yx-0 cZGbLO"
    data-testid="dropdown_id"
    id="dropdown_id"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Dropdown Menu"
      className="dropdown-toggle"
      data-testid="dropdown_id_toggle"
      id="dropdown_id_toggle"
      onClick={[Function]}
    >
      Toggle Dropdown
    </button>
    <ul
      className="sc-2bi9jj-1 gEomZU dropdown-menu "
      data-testid="dropdown_id_list"
      id="dropdown_id_list"
    >
      <li
        className=" "
        data-testid="dropdown_id_list_0"
        id="dropdown_id_list_0"
        onClick={[Function]}
      >
        <a
          href="#"
        >
          Im a Link
        </a>
      </li>
      <li
        className=" "
        data-testid="dropdown_id_list_1"
        id="dropdown_id_list_1"
        onClick={[Function]}
      >
        Item Two
      </li>
      <li
        className=" "
        data-testid="dropdown_id_list_2"
        id="dropdown_id_list_2"
        onClick={[Function]}
      >
        Item Three
      </li>
    </ul>
  </nav>
  <nav
    className="sc-sg24yx-0 heFPof"
    data-testid="dropdown_id"
    id="dropdown_id"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Dropdown Menu"
      className="dropdown-toggle"
      data-testid="dropdown_id_toggle"
      id="dropdown_id_toggle"
      onClick={[Function]}
    >
      Toggle Dropdown
    </button>
    <ul
      className="sc-2bi9jj-1 gEomZU dropdown-menu "
      data-testid="dropdown_id_list"
      id="dropdown_id_list"
    >
      <li
        className=" "
        data-testid="dropdown_id_list_0"
        id="dropdown_id_list_0"
        onClick={[Function]}
      >
        <a
          href="#"
        >
          Im a Link
        </a>
      </li>
      <li
        className=" "
        data-testid="dropdown_id_list_1"
        id="dropdown_id_list_1"
        onClick={[Function]}
      >
        Item Two
      </li>
      <li
        className=" "
        data-testid="dropdown_id_list_2"
        id="dropdown_id_list_2"
        onClick={[Function]}
      >
        Item Three
      </li>
    </ul>
  </nav>
  <nav
    className="sc-sg24yx-0 lnjURR"
    data-testid="dropdown_id"
    id="dropdown_id"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Dropdown Menu"
      className="dropdown-toggle"
      data-testid="dropdown_id_toggle"
      id="dropdown_id_toggle"
      onClick={[Function]}
    >
      Toggle Dropdown
    </button>
    <ul
      className="sc-2bi9jj-1 gEomZU dropdown-menu "
      data-testid="dropdown_id_list"
      id="dropdown_id_list"
    >
      <li
        className=" "
        data-testid="dropdown_id_list_0"
        id="dropdown_id_list_0"
        onClick={[Function]}
      >
        <a
          href="#"
        >
          Im a Link
        </a>
      </li>
      <li
        className=" "
        data-testid="dropdown_id_list_1"
        id="dropdown_id_list_1"
        onClick={[Function]}
      >
        Item Two
      </li>
      <li
        className=" "
        data-testid="dropdown_id_list_2"
        id="dropdown_id_list_2"
        onClick={[Function]}
      >
        Item Three
      </li>
    </ul>
  </nav>
</div>
`;

exports[`Storyshots Molecules/Dropdown Default 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <nav
    className="sc-sg24yx-0 heFPof"
    data-testid="dropdown_id"
    id="dropdown_id"
  >
    <button
      aria-expanded="false"
      aria-haspopup="true"
      aria-label="Dropdown Menu"
      className="dropdown-toggle"
      data-testid="dropdown_id_toggle"
      id="dropdown_id_toggle"
      onClick={[Function]}
    >
      Toggle Dropdown
    </button>
    <ul
      className="sc-2bi9jj-1 gEomZU dropdown-menu "
      data-testid="dropdown_id_list"
      id="dropdown_id_list"
    >
      <li
        className=" "
        data-testid="dropdown_id_list_0"
        id="dropdown_id_list_0"
        onClick={[Function]}
      >
        <a
          href="#"
        >
          Im a Link
        </a>
      </li>
      <li
        className=" "
        data-testid="dropdown_id_list_1"
        id="dropdown_id_list_1"
        onClick={[Function]}
      >
        Item Two
      </li>
      <li
        className=" "
        data-testid="dropdown_id_list_2"
        id="dropdown_id_list_2"
        onClick={[Function]}
      >
        Item Three
      </li>
    </ul>
  </nav>
</div>
`;

exports[`Storyshots Molecules/Dropdown Skeleton 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-kjnynt-0 jfZhXD"
    data-testid="dropdown_id_skeleton"
    id="dropdown_id_skeleton"
    style={
      Object {
        "height": undefined,
        "width": undefined,
      }
    }
  >
    <nav
      className="sc-sg24yx-0 heFPof"
      data-testid="dropdown_id"
      id="dropdown_id"
    >
      <button
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="Dropdown Menu"
        className="dropdown-toggle"
        data-testid="dropdown_id_toggle"
        id="dropdown_id_toggle"
        onClick={[Function]}
      >
        Toggle Dropdown
      </button>
      <ul
        className="sc-2bi9jj-1 gEomZU dropdown-menu "
        data-testid="dropdown_id_list"
        id="dropdown_id_list"
      >
        <li
          className=" "
          data-testid="dropdown_id_list_0"
          id="dropdown_id_list_0"
          onClick={[Function]}
        >
          <a
            href="#"
          >
            Im a Link
          </a>
        </li>
        <li
          className=" "
          data-testid="dropdown_id_list_1"
          id="dropdown_id_list_1"
          onClick={[Function]}
        >
          Item Two
        </li>
        <li
          className=" "
          data-testid="dropdown_id_list_2"
          id="dropdown_id_list_2"
          onClick={[Function]}
        >
          Item Three
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
